### ISSUE TITLE
Rust attribute changes are not ideally associated

### ISSUE THREAD

**Issue by cuviper** [Wilfred/difftastic#35]
Using the [indexmap](https://github.com/bluss/indexmap) repo, I tried the following command:

    GIT_EXTERNAL_DIFF=difftastic git show --ext-diff 49d430808ced6d1824ec15d621acd6b0e8367c19

and got this output near the end:

![image](https://user-images.githubusercontent.com/36186/133663966-865f16c0-9283-48ae-877d-43266786be57.png)

Notice how the `#[test]` attribute is interpreted, where the following one is considered to be an insertion, instead of the one that's on the new function. This kind of thing is very common with line-based diffs, although plain `git show` actually does get this right!

![image](https://user-images.githubusercontent.com/36186/133664758-2b4c1e37-28c2-432d-b989-962f146fd636.png)

**Comment by Wilfred** [comment]
Yeah, this is sometimes called a "slider", and difftastic isn't completely immune.

I have been thinking about trying to nudge the diff logic to avoid blank lines, but it might increase complexity. The best solution for Rust here is probably to improve the tree-sitter parser to include the attribute with the function definition.

**Comment by gthb** [comment]
I'm seeing the exact slider behavior described above, in Difftastic 0.31.0, in my codebase (redacting the test function body, private and probably not relevant):

![image](https://user-images.githubusercontent.com/153580/180213450-f96ecd1a-2c91-4785-bb81-134c4c6ef9f9.png)

However, it does get the above `indexmap` example right:

![image](https://user-images.githubusercontent.com/153580/180213031-0926c269-2eb9-4bfe-87a9-de44b2eb903a.png)

Not sure what the difference is. Maybe these just fall on either side of some heuristic threshold?

**Comment by wetneb** [comment]
I have improved the Rust parser to fix the attachment of attributes to functions (and many types of nodes).
The result is available in a fork called [`tree-sitter-rust-orchard`](https://codeberg.org/grammar-orchard/tree-sitter-rust-orchard), which is [available on crates.io](https://crates.io/crates/tree-sitter-rust-orchard). The reason why it's in a fork is that PRs to tree-sitter-rust don't get reviewed very often.

This fork contains many other improvements which add support for newer language constructs (including unstable ones). See [the changelog](https://codeberg.org/grammar-orchard/tree-sitter-rust-orchard/releases) for details about those improvements.
